composer/algorithms/alibi/attention_surgery_functions/_bert.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+from types import MethodType
+from typing import Optional, Tuple
+
+import torch
+from torch import nn
+from transformers.models.bert.modeling_bert import BertEmbeddings, BertSelfAttention
+from transformers.models.roberta.modeling_roberta import RobertaEmbeddings, RobertaSelfAttention
+
+from composer.algorithms.alibi.attention_surgery_functions.utils import (ReplacementFunction, register_alibi,
+                                                                         register_surgery_function_builder,
+                                                                         zero_and_freeze_expand_position_embeddings)
+
+
+@register_surgery_function_builder(BertEmbeddings, RobertaEmbeddings)
+def build_bert_embedding_converter(max_sequence_length: int) -> ReplacementFunction:
+    """Builds a function to remove positional embeddings.
+
+    The built function also expands `position_ids` buffer to support `max_sequence_length` tokens.
+    """
+
+    def convert_position_embeddings_and_ids(module: torch.nn.Module, module_index: Optional[int] = None):
+        assert isinstance(module, (BertEmbeddings, RobertaEmbeddings))
+        del module_index  # unused
+        zero_and_freeze_expand_position_embeddings(module,
+                                                   max_sequence_length,
+                                                   position_embedding_attribute='position_embeddings')
+
+        module_device = next(module.parameters()).device
+        module.register_buffer('position_ids', torch.arange(max_sequence_length).expand((1, -1)).to(module_device))
+        return module
+
+    return convert_position_embeddings_and_ids
+
+
+@register_surgery_function_builder(BertSelfAttention, RobertaSelfAttention)
+def build_bert_attention_converter(max_sequence_length: int) -> ReplacementFunction:
+    """Builds a function that does model surgery to add ALiBi to Bert-style SelfAttention."""
+
+    def convert_attention(module: torch.nn.Module, module_index: Optional[int] = None):
+        assert isinstance(module, (BertSelfAttention, RobertaSelfAttention))
+        del module_index  # unused
+        module = register_alibi(module=module,
+                                n_heads=int(module.num_attention_heads),
+                                max_token_length=max_sequence_length,
+                                causal=False)
+        setattr(module, 'forward', MethodType(forward, module))
+
+        return module
+
+    return convert_attention
+
+
+# pyright: reportGeneralTypeIssues = none
```

Comment:
Is there a reason we decided to override the forward function rather than replace the class instances? (cc @hanlint , I think he wrote the same comment and reached out offline)

Reply:
A few reasons. They're listed out in a separate comment below. Copied here:

>> I did a first pass and it looks good to me so far, but why do we override the forward method instead of replacing all instances of a particular attention mechanism? (cc @hanlint -- thoughts on the design?)

> @moinnadeem Great question. This was my plan early on but I switched gears. There are a few main reasons.
> 
> Replacing modules with new module instances would require additional boilerplate code to define the replacement class, which is very likely going to be only slightly different than the class type it replaces during surgery.
> It is not necessarily trivial to know what parameter values to pass when instantiating the replacement class. By modifying the module instance rather than replacing it you avoid the problem of having to correctly initializing each new instance.
> Overriding the module function (rather than replacing the module instance itself) means that module surgery leaves the module's parameter weights intact. This is desirable, e.g., if the pre-surgery module's parameters are already trained -- for instance, if the module weights had been loaded in from some checkpoint.